Flexible handling of dtstart incl. documentation

denisiko opened the pull request:
As stated by @jpulec in #90 `dtstart` is handled according to RFC 2445 so that it always counts as the first occurrence regardless of your pattern, like when used with `between()`. This causes confusion for some users (e.g. #36, #50, #71) because it is in conflict with how `dateutil.rrule` handles `dtstart`.

With this pull request I suggest a solution which consists of:

1. The addition of an optional parameter `include_dtstart` which can be either passed to a `RecurrenceField` or a `Recurrence` instance. With `include_dtstart=False` the default behavior does not occur for the field/object so that `dtstart` no longer finds its way into the occurrence list if it does not match your recurrence pattern. Because the default behavior doesn't change if `include_dtstart` is not specified backward compatibility is ensured.

2. A clear documentation of how `django-recurrence` handles `dtstart` and how it can be controlled with the feature described above. In that way further confusion should be prevented.

3. Some additional unit test cases.

fozzle commented:
This is fantastic, thank you. The current behavior is very unintuitive but I understand it's part of the spec. This would solve a lot of problems let's get this in!

jpulec commented:
@emperorcezar Would it be possible to get this looked at?

emperorcezar commented:
@denisiko What's here looks good to me with the exception of two things. I would like to see some tests, and second, adding it to the release notes.

Get those in and I think we'd be good to go.

denisiko commented:
@emperorcezar Thanks for your feedback!  I will write some test cases when I'm back from holidays at the end of the month.

denisiko commented:
@emperorcezar So I finally managed to add some tests. They pick up on the example mentioned in #50.

emperorcezar commented:
@denisiko Awesome.
@dominicrodger Could you take a quick look.

denisiko commented:
@emperorcezar I guess he's busy. Would you mind taking a look?

emperorcezar commented:
@denisiko Code looks good. Could I suggest the flag be `include_dtstart`? I think that's much more intuitive. `inc` makes my mind go straight to increment.

denisiko commented:
@emperorcezar Thx for the very quick reply!
Yes, that's a very good suggestion! I refactored the flag everywhere and updated this request's description.

denisiko commented:
@emperorcezar Is there anything else I can do?

emperorcezar commented:
@denisiko I was hoping for @dominicrodger to take a look. I mostly do release management.

denisiko commented:
I totally get that. But this request lies around for half a year now...

I don't mind waiting longer because I do not encounter a problem with dtstart in my project. I adressed this problem because a lot of people are affected and wait for a solution.

dominicrodger commented:
Sorry folks - I added a third child to my little clan a few months ago, so spare time has been hard to come by lately! I'll take a look at it this week. If that deadline passes, feel free to remind me (it's in my inbox now, so I shouldn't forget).

Thanks for the work and patience here - it's much appreciated.

denisiko commented:
Wow congrats @dominicrodger ! Well I can imagine then that you have other more urgent things on your mind.

@emperorcezar already mentioned some good points while looking over it, so I hope there are not a lot problems left. But I am very much looking forward to your feedback on the approach I toke.

denisiko commented:
@jpulec As you may have seen I did not include the param in the functions, so call by call overrides are not supported e.g. with between.

I just don't think there is a use case where someone wants both behaviors in one app: so either you want dtstart included everywhere, you don't want it anywhere or you just don't care (e.g when you don't work with past dates). But even if you would want that you could set include_dtstart to False and just add dtstart explicitly with a one-liner wherever you want it included.

dominicrodger commented:
This looks good to me - thanks so much for your work here @denisiko!

Only thing left I think is to update the changelog in a few places, but I'll get that handled, and then we can push a release. I'll update this ticket with news on that, most likely tomorrow.

denisiko commented:
No problem, glad I could help.

denisiko commented:
@dominicrodger 
Can you already say when the next pip release is going to be? After that some issues could be closed.

emperorcezar commented:
I can push a new release whenever I get the go-ahead.

On Tue, Oct 10, 2017, 4:51 AM Denis Anuschewski <notifications@github.com>
wrote:

> @dominicrodger <https://github.com/dominicrodger>
> Can you already say when the next pip release is going to be? After that
> some issues could be closed.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/django-recurrence/django-recurrence/pull/93#issuecomment-335421430>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAELN0HvTn9YAWEDxz6zqk-T6NlIpp6Uks5sqz4pgaJpZM4NHobn>
> .
>

dominicrodger commented:
Go ahead :)

emperorcezar commented:
Pushed.

On Tue, Oct 10, 2017 at 8:39 AM Dominic Rodger <notifications@github.com>
wrote:

> Go ahead :)
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/django-recurrence/django-recurrence/pull/93#issuecomment-335475592>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAELN5LvH5Qu25qcwJLDE5ICXq1tpchAks5sq3OqgaJpZM4NHobn>
> .
>
